### ISSUE TITLE
array_index does not catch std::invalid_argument exception from std::stoull

### ISSUE THREAD

**Issue by StarryVae** [nlohmann/json#3831]
### Description

array_index does not catch `std::invalid_argument` exception from `std::stoull`, which means we need catch the exception in our code when using nlohmann json.

for example, for a json body `json_body`:  `{\"foo\":\"val\",\"bar\":\"val\",\"array\":[\"a1\",\"a2\"]}`, if json pointer `json_pointer` was set incorrectly like `/array/`, when we call `json_body.contains(json_pointer)`, it will throw `std::invalid_argument` exception.

### Reproduction steps

for a json body `json_body`:  `{\"foo\":\"val\",\"bar\":\"val\",\"array\":[\"a1\",\"a2\"]}`, if json pointer `json_pointer` was set incorrectly like `/array/`, when we call `json_body.contains(json_pointer)`, it will throw `std::invalid_argument` exception.

### Expected vs. actual results

Expected:
catch `nlohmann::detail::exception` is enough.

Actual:
just catch `nlohmann::detail::exception` is not enough, crush will happen due to `std::invalid_argument` exception.

### Compiler and operating system

clang 7.0.1 ubuntu 18.04

### Library version

3.10.5

### Validation

- [ ] The bug also occurs if the latest version from the [`develop`](https://github.com/nlohmann/json/tree/develop) branch is used.
- [ ] I can successfully [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests).

**Comment by falbrechtskirchinger** [comment]
This has been fixed (#3415).

v3.10.5
```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  stoull
```

v3.11.2
```
terminate called after throwing an instance of 'nlohmann::json_abi_v3_11_2::detail::out_of_range'
  what():  [json.exception.out_of_range.404] unresolved reference token ''
```

**Comment by StarryVae** [comment]
> This has been fixed (#3415).
> 
> v3.10.5
> 
> ```
> terminate called after throwing an instance of 'std::invalid_argument'
>   what():  stoull
> ```
> 
> v3.11.2
> 
> ```
> terminate called after throwing an instance of 'nlohmann::json_abi_v3_11_2::detail::out_of_range'
>   what():  [json.exception.out_of_range.404] unresolved reference token ''
> ```

thanks!
